packages/sql-events/index.js

```diff
@@ -0,0 +1,140 @@
+'use strict'
+
+const MQEmitter = require('mqemitter')
+const fp = require('fastify-plugin')
+const { PassThrough } = require('stream')
+const MQEmitterRedis = require('mqemitter-redis')
+const { promisify } = require('util')
+
+async function fastifySqlEvents (app, opts) {
+  setupEmitter({ ...opts, mapper: app.platformatic, log: app.log })
+  const mq = app.platformatic.mq
+  app.addHook('onClose', () => promisify(mq.close.bind(mq)))
+}
+
+function setupEmitter ({ log, mq, mapper, connectionString }) {
+  if (connectionString) {
+    mq = MQEmitterRedis({ connectionString })
+  } else if (!mq) {
+    mq = MQEmitter()
+  }
+
+  for (const entityName of Object.keys(mapper.entities)) {
+    const entity = mapper.entities[entityName]
+    const { primaryKey } = entity
+    mapper.addEntityHooks(entityName, {
+      async save (original, data) {
+        const ctx = data.ctx
+        /* istanbul ignore next */
+        const _log = ctx?.reply?.request?.log || log
+        const res = await original(data)
+        const topic = await entity.getPublishTopic({ action: 'save', data: res, ctx })
+        if (topic) {
+          const payload = {
+            [primaryKey]: res[primaryKey]
+          }
+          _log.trace({ topic, payload }, 'publishing event')
+          await new Promise((resolve) => {
+            mq.emit({
+              topic,
+              payload
+            }, resolve)
+          })
+        }
+        return res
+      },
+
+      delete: multiField('delete'),
+      insert: multiField('save')
+    })
+
+    function multiField (action) {
+      return async function (original, data) {
+        const ctx = data.ctx
+        /* istanbul ignore next */
+        const _log = ctx?.reply?.request?.log || log
+        const fields = new Set(data.fields)
+        const res = await original({ ...data, fields: undefined })
+
+        await Promise.all(res.map(async (payload) => {
+          const topic = await entity.getPublishTopic({ action, data: payload, ctx })
+          if (topic) {
+            _log.trace({ topic, payload }, 'publishing event')
+            return new Promise((resolve) => {
+              mq.emit({
+                topic,
+                payload
+              }, resolve)
+            })
+          }
+        }))
+
+        if (fields.size > 0) {
+          const actual = []
+          for (const element of res) {
+            const obj = {}
+            for (const field of fields) {
+              obj[field] = element[field]
+            }
+            actual.push(obj)
+          }
+          return actual
+        } else {
+          return res
+        }
+      }
+    }
+
+    entity.getPublishTopic = async function ({ action, data }) {
```

Comment:
The only field you need from the `data` argument is a primary key. You can receive it directly.

Reply:
I'm aiming to provide the flexibility of using other attributes as well - if folks override these via hooks.